File: backend/tax/engine.py
"""Core tax computation engine for Indian income tax."""

from .slabs import (
    NEW_REGIME_SLABS, OLD_REGIME_SLABS,
    REBATE_87A_NEW, REBATE_87A_OLD,
    SURCHARGE_RATES, CESS_RATE,
    PRESUMPTIVE_44ADA, STANDARD_DEDUCTION_NEW,
)
from .deductions import compute_deductions


def compute_tax_on_slabs(taxable_income: float, slabs: list[tuple]) -> float:
    """Compute tax using given slab rates."""
    tax = 0.0
    remaining = taxable_income

    for slab_limit, rate in slabs:
        if slab_limit is None:
            # Last slab, no upper limit
            tax += remaining * rate
            break
        taxable_in_slab = min(remaining, slab_limit)
        tax += taxable_in_slab * rate
        remaining -= taxable_in_slab
        if remaining <= 0:
            break

    return round(tax, 2)


def compute_surcharge(tax: float, taxable_income: float) -> float:
    """Compute surcharge on income tax with marginal relief."""
    applicable_rate = 0.0
    for threshold, rate in SURCHARGE_RATES:
        if threshold is None or taxable_income <= threshold:
            applicable_rate = rate
            break
        applicable_rate = rate

    surcharge = round(tax * applicable_rate, 2)

    # Marginal relief: surcharge should not exceed income exceeding the threshold
    if applicable_rate > 0:
        prev_threshold = 0
        for threshold, rate in SURCHARGE_RATES:
            if rate == applicable_rate and prev_threshold > 0:
                excess_income = taxable_income - prev_threshold
                tax_at_prev = compute_tax_on_slabs(prev_threshold, NEW_REGIME_SLABS)
                if tax + surcharge > tax_at_prev + excess_income:
                    surcharge = max(0, (tax_at_prev + excess_income) - tax)
                break
            if threshold is not None:
                prev_threshold = threshold

    return round(surcharge, 2)


def compute_cess(tax_plus_surcharge: float) -> float:
    """4% Health and Education Cess."""
    return round(tax_plus_surcharge * CESS_RATE, 2)


def compute_rebate_87a(tax: float, taxable_income: float, regime: str) -> float:
    """Compute rebate under Section 87A."""
    if regime == "new":
        if taxable_income <= REBATE_87A_NEW["limit"]:
            return min(tax, REBATE_87A_NEW["max_rebate"])
    else:
        if taxable_income <= REBATE_87A_OLD["limit"]:
            return min(tax, REBATE_87A_OLD["max_rebate"])
    return 0.0


def compute_itr4(
    gross_professional_receipts: float,
    other_income: dict,
    deductions_list: list[dict],
    tds_credit: float = 0.0,
    advance_tax_paid: float = 0.0,
    regime: str = "new",
) -> dict:
    """
    Compute tax for ITR-4 (Presumptive taxation under 44ADA).
    50% of gross receipts deemed as income.
    """
    deemed_profit_rate = PRESUMPTIVE_44ADA["deemed_profit_rate"]
    deemed_income = round(gross_professional_receipts * deemed_profit_rate, 2)

    interest = other_income.get("interest", 0.0)
    rental = other_income.get("rental", 0.0)
    dividend = other_income.get("dividend", 0.0)
    other = other_income.get("other", 0.0)
    gross_other = interest + rental + dividend + other

    gross_total_income = deemed_income + gross_other

    if regime == "old":
        total_deductions = compute_deductions(deductions_list)
    else:
        # New regime: standard deduction of 75K for salaried only
        # For business/professional income under 44ADA, no standard deduction
        total_deductions = 0.0

    taxable_income = max(gross_total_income - total_deductions, 0.0)

    slabs = NEW_REGIME_SLABS if regime == "new" else OLD_REGIME_SLABS
    tax_on_income = compute_tax_on_slabs(taxable_income, slabs)

    rebate = compute_rebate_87a(tax_on_income, taxable_income, regime)
    tax_after_rebate = max(tax_on_income - rebate, 0.0)

    surcharge = compute_surcharge(tax_after_rebate, taxable_income)
    cess = compute_cess(tax_after_rebate + surcharge)
    total_tax = round(tax_after_rebate + surcharge + cess, 2)

    net_tax = round(total_tax - tds_credit - advance_tax_paid, 2)

    return {
        "fy": "2025-26",
        "regime": regime,
        "itr_form": "ITR-4",
        "gross_professional_income": gross_professional_receipts,
        "deemed_income": deemed_income,
        "gross_other_income": gross_other,
        "gross_total_income": gross_total_income,
        "business_expenses": None,
        "total_deductions": total_deductions,
        "taxable_income": taxable_income,
        "tax_on_income": tax_on_income,
        "rebate_87a": rebate,
        "surcharge": surcharge,
        "cess": cess,
        "total_tax": total_tax,
        "tds_credit": tds_credit,
        "advance_tax_paid": advance_tax_paid,
        "tax_payable": max(net_tax, 0.0),
        "tax_refund": abs(min(net_tax, 0.0)),
    }


def compute_itr3(
    gross_professional_income: float,
    business_expenses: float,
    other_income: dict,
    deductions_list: list[dict],
    tds_credit: float = 0.0,
    advance_tax_paid: float = 0.0,
    regime: str = "new",
) -> dict:
    """
    Compute tax for ITR-3 (Business/Profession with books of accounts).
    Income = Receipts - Expenses.
    """
    net_business_income = max(gross_professional_income - business_expenses, 0.0)

    interest = other_income.get("interest", 0.0)
    rental = other_income.get("rental", 0.0)
    dividend = other_income.get("dividend", 0.0)
    other = other_income.get("other", 0.0)
    gross_other = interest + rental + dividend + other

    gross_total_income = net_business_income + gross_other

    if regime == "old":
        total_deductions = compute_deductions(deductions_list)
    else:
        total_deductions = 0.0

    taxable_income = max(gross_total_income - total_deductions, 0.0)

    slabs = NEW_REGIME_SLABS if regime == "new" else OLD_REGIME_SLABS
    tax_on_income = compute_tax_on_slabs(taxable_income, slabs)

    rebate = compute_rebate_87a(tax_on_income, taxable_income, regime)
    tax_after_rebate = max(tax_on_income - rebate, 0.0)

    surcharge = compute_surcharge(tax_after_rebate, taxable_income)
    cess = compute_cess(tax_after_rebate + surcharge)
    total_tax = round(tax_after_rebate + surcharge + cess, 2)

    net_tax = round(total_tax - tds_credit - advance_tax_paid, 2)

    return {
        "fy": "2025-26",
        "regime": regime,
        "itr_form": "ITR-3",
        "gross_professional_income": gross_professional_income,
        "deemed_income": None,
        "business_expenses": business_expenses,
        "gross_other_income": gross_other,
        "gross_total_income": gross_total_income,
        "total_deductions": total_deductions,
        "taxable_income": taxable_income,
        "tax_on_income": tax_on_income,
        "rebate_87a": rebate,
        "surcharge": surcharge,
        "cess": cess,
        "total_tax": total_tax,
        "tds_credit": tds_credit,
        "advance_tax_paid": advance_tax_paid,
        "tax_payable": max(net_tax, 0.0),
        "tax_refund": abs(min(net_tax, 0.0)),
    }


def compare_regimes(
    gross_professional_income: float,
    business_expenses: float,
    other_income: dict,
    deductions_list: list[dict],
    tds_credit: float = 0.0,
    advance_tax_paid: float = 0.0,
    itr_form: str = "ITR-4",
) -> dict:
    """Compare old vs new regime and recommend the better one."""
    compute_fn = compute_itr4 if itr_form == "ITR-4" else compute_itr3

    if itr_form == "ITR-4":
        old = compute_fn(gross_professional_income, other_income, deductions_list, tds_credit, advance_tax_paid, "old")
        new = compute_fn(gross_professional_income, other_income, deductions_list, tds_credit, advance_tax_paid, "new")
    else:
        old = compute_fn(gross_professional_income, business_expenses, other_income, deductions_list, tds_credit, advance_tax_paid, "old")
        new = compute_fn(gross_professional_income, business_expenses, other_income, deductions_list, tds_credit, advance_tax_paid, "new")

    old_payable = old["tax_payable"] - old["tax_refund"]
    new_payable = new["tax_payable"] - new["tax_refund"]

    recommended = "new" if new_payable <= old_payable else "old"
    savings = abs(old_payable - new_payable)

    return {
        "old_regime": old,
        "new_regime": new,
        "recommended": recommended,
        "savings": round(savings, 2),
    }

File: frontend/src/lib/tax/engine.ts
/**
 * Core tax computation engine for Indian income tax.
 */

import type { TaxResult, RegimeComparison } from "@/lib/types";
import {
  NEW_REGIME_SLABS,
  OLD_REGIME_SLABS,
  REBATE_87A_NEW,
  REBATE_87A_OLD,
  SURCHARGE_RATES,
  CESS_RATE,
  PRESUMPTIVE_44ADA,
  type Slab,
} from "./slabs";
import { computeDeductions } from "./deductions";

function round2(x: number): number {
  return Math.round(x * 100) / 100;
}

interface OtherIncome {
  interest?: number;
  rental?: number;
  dividend?: number;
  other?: number;
}

interface DeductionInput {
  section?: string;
  amount?: number;
  [key: string]: unknown;
}

/** Compute tax using given slab rates. */
export function computeTaxOnSlabs(
  taxableIncome: number,
  slabs: Slab[]
): number {
  let tax = 0;
  let remaining = taxableIncome;

  for (const [slabLimit, rate] of slabs) {
    if (slabLimit === null) {
      // Last slab, no upper limit
      tax += remaining * rate;
      break;
    }
    const taxableInSlab = Math.min(remaining, slabLimit);
    tax += taxableInSlab * rate;
    remaining -= taxableInSlab;
    if (remaining <= 0) {
      break;
    }
  }

  return round2(tax);
}

/** Compute surcharge on income tax with marginal relief. */
export function computeSurcharge(
  tax: number,
  taxableIncome: number
): number {
  let applicableRate = 0;
  for (const [threshold, rate] of SURCHARGE_RATES) {
    if (threshold === null || taxableIncome <= threshold) {
      applicableRate = rate;
      break;
    }
    applicableRate = rate;
  }

  let surcharge = round2(tax * applicableRate);

  // Marginal relief: surcharge should not exceed income exceeding the threshold
  if (applicableRate > 0) {
    let prevThreshold = 0;
    for (const [threshold, rate] of SURCHARGE_RATES) {
      if (rate === applicableRate && prevThreshold > 0) {
        const excessIncome = taxableIncome - prevThreshold;
        const taxAtPrev = computeTaxOnSlabs(prevThreshold, NEW_REGIME_SLABS);
        if (tax + surcharge > taxAtPrev + excessIncome) {
          surcharge = Math.max(0, taxAtPrev + excessIncome - tax);
        }
        break;
      }
      if (threshold !== null) {
        prevThreshold = threshold;
      }
    }
  }

  return round2(surcharge);
}

/** 4% Health and Education Cess. */
export function computeCess(taxPlusSurcharge: number): number {
  return round2(taxPlusSurcharge * CESS_RATE);
}

/** Compute rebate under Section 87A. */
export function computeRebate87a(
  tax: number,
  taxableIncome: number,
  regime: string
): number {
  if (regime === "new") {
    if (taxableIncome <= REBATE_87A_NEW.limit) {
      return Math.min(tax, REBATE_87A_NEW.max_rebate);
    }
  } else {
    if (taxableIncome <= REBATE_87A_OLD.limit) {
      return Math.min(tax, REBATE_87A_OLD.max_rebate);
    }
  }
  return 0;
}

/**
 * Compute tax for ITR-4 (Presumptive taxation under 44ADA).
 * 50% of gross receipts deemed as income.
 */
export function computeItr4(
  grossProfessionalReceipts: number,
  otherIncome: OtherIncome,
  deductionsList: DeductionInput[],
  tdsCredit: number = 0,
  advanceTaxPaid: number = 0,
  regime: string = "new"
): TaxResult {
  const deemedProfitRate = PRESUMPTIVE_44ADA.deemed_profit_rate;
  const deemedIncome = round2(grossProfessionalReceipts * deemedProfitRate);

  const interest = otherIncome.interest ?? 0;
  const rental = otherIncome.rental ?? 0;
  const dividend = otherIncome.dividend ?? 0;
  const other = otherIncome.other ?? 0;
  const grossOther = interest + rental + dividend + other;

  const grossTotalIncome = deemedIncome + grossOther;

  let totalDeductions: number;
  if (regime === "old") {
    totalDeductions = computeDeductions(deductionsList);
  } else {
    // New regime: no standard deduction for business/professional income under 44ADA
    totalDeductions = 0;
  }

  const taxableIncome = Math.max(grossTotalIncome - totalDeductions, 0);

  const slabs = regime === "new" ? NEW_REGIME_SLABS : OLD_REGIME_SLABS;
  const taxOnIncome = computeTaxOnSlabs(taxableIncome, slabs);

  const rebate = computeRebate87a(taxOnIncome, taxableIncome, regime);
  const taxAfterRebate = Math.max(taxOnIncome - rebate, 0);

  const surcharge = computeSurcharge(taxAfterRebate, taxableIncome);
  const cess = computeCess(taxAfterRebate + surcharge);
  const totalTax = round2(taxAfterRebate + surcharge + cess);

  const netTax = round2(totalTax - tdsCredit - advanceTaxPaid);

  return {
    fy: "2025-26",
    regime,
    itr_form: "ITR-4",
    gross_professional_income: grossProfessionalReceipts,
    deemed_income: deemedIncome,
    business_expenses: null,
    gross_other_income: grossOther,
    gross_total_income: grossTotalIncome,
    total_deductions: totalDeductions,
    taxable_income: taxableIncome,
    tax_on_income: taxOnIncome,
    rebate_87a: rebate,
    surcharge,
    cess,
    total_tax: totalTax,
    tds_credit: tdsCredit,
    advance_tax_paid: advanceTaxPaid,
    tax_payable: Math.max(netTax, 0),
    tax_refund: Math.abs(Math.min(netTax, 0)),
  };
}

/**
 * Compute tax for ITR-3 (Business/Profession with books of accounts).
 * Income = Receipts - Expenses.
 */
export function computeItr3(
  grossProfessionalIncome: number,
  businessExpenses: number,
  otherIncome: OtherIncome,
  deductionsList: DeductionInput[],
  tdsCredit: number = 0,
  advanceTaxPaid: number = 0,
  regime: string = "new"
): TaxResult {
  const netBusinessIncome = Math.max(
    grossProfessionalIncome - businessExpenses,
    0
  );

  const interest = otherIncome.interest ?? 0;
  const rental = otherIncome.rental ?? 0;
  const dividend = otherIncome.dividend ?? 0;
  const other = otherIncome.other ?? 0;
  const grossOther = interest + rental + dividend + other;

  const grossTotalIncome = netBusinessIncome + grossOther;

  let totalDeductions: number;
  if (regime === "old") {
    totalDeductions = computeDeductions(deductionsList);
  } else {
    totalDeductions = 0;
  }

  const taxableIncome = Math.max(grossTotalIncome - totalDeductions, 0);

  const slabs = regime === "new" ? NEW_REGIME_SLABS : OLD_REGIME_SLABS;
  const taxOnIncome = computeTaxOnSlabs(taxableIncome, slabs);

  const rebate = computeRebate87a(taxOnIncome, taxableIncome, regime);
  const taxAfterRebate = Math.max(taxOnIncome - rebate, 0);

  const surcharge = computeSurcharge(taxAfterRebate, taxableIncome);
  const cess = computeCess(taxAfterRebate + surcharge);
  const totalTax = round2(taxAfterRebate + surcharge + cess);

  const netTax = round2(totalTax - tdsCredit - advanceTaxPaid);

  return {
    fy: "2025-26",
    regime,
    itr_form: "ITR-3",
    gross_professional_income: grossProfessionalIncome,
    deemed_income: null,
    business_expenses: businessExpenses,
    gross_other_income: grossOther,
    gross_total_income: grossTotalIncome,
    total_deductions: totalDeductions,
    taxable_income: taxableIncome,
    tax_on_income: taxOnIncome,
    rebate_87a: rebate,
    surcharge,
    cess,
    total_tax: totalTax,
    tds_credit: tdsCredit,
    advance_tax_paid: advanceTaxPaid,
    tax_payable: Math.max(netTax, 0),
    tax_refund: Math.abs(Math.min(netTax, 0)),
  };
}

/** Compare old vs new regime and recommend the better one. */
export function compareRegimes(
  grossProfessionalIncome: number,
  businessExpenses: number,
  otherIncome: OtherIncome,
  deductionsList: DeductionInput[],
  tdsCredit: number = 0,
  advanceTaxPaid: number = 0,
  itrForm: string = "ITR-4"
): RegimeComparison {
  let oldResult: TaxResult;
  let newResult: TaxResult;

  if (itrForm === "ITR-4") {
    oldResult = computeItr4(
      grossProfessionalIncome,
      otherIncome,
      deductionsList,
      tdsCredit,
      advanceTaxPaid,
      "old"
    );
    newResult = computeItr4(
      grossProfessionalIncome,
      otherIncome,
      deductionsList,
      tdsCredit,
      advanceTaxPaid,
      "new"
    );
  } else {
    oldResult = computeItr3(
      grossProfessionalIncome,
      businessExpenses,
      otherIncome,
      deductionsList,
      tdsCredit,
      advanceTaxPaid,
      "old"
    );
    newResult = computeItr3(
      grossProfessionalIncome,
      businessExpenses,
      otherIncome,
      deductionsList,
      tdsCredit,
      advanceTaxPaid,
      "new"
    );
  }

  const oldPayable = oldResult.tax_payable - oldResult.tax_refund;
  const newPayable = newResult.tax_payable - newResult.tax_refund;

  const recommended = newPayable <= oldPayable ? "new" : "old";
  const savings = Math.abs(oldPayable - newPayable);

  return {
    old_regime: oldResult,
    new_regime: newResult,
    recommended,
    savings: round2(savings),
  };
}
